From 99ce927dd4f10c194f93a7a68c35e4a685bf3077 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Sat, 10 Feb 2018 10:56:56 +0100
Subject: [PATCH] fsl_spdif: Do not keep references to unused clocks

In order to find the best clock source, the rate matching code may
acquire several clock instances. But it does not release them in
case they are not chosen.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 sound/soc/fsl/fsl_spdif.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/sound/soc/fsl/fsl_spdif.c b/sound/soc/fsl/fsl_spdif.c
index e712c03..9b10415 100644
--- a/sound/soc/fsl/fsl_spdif.c
+++ b/sound/soc/fsl/fsl_spdif.c
@@ -1193,13 +1193,20 @@ static int fsl_spdif_probe_txclk(struct fsl_spdif_priv *spdif_priv,
 			dev_err(dev, "no rxtx%d clock in devicetree\n", i);
 			return PTR_ERR(clk);
 		}
-		if (!clk_get_rate(clk))
+		if (!clk_get_rate(clk)) {
+			devm_clk_put(&pdev->dev, clk);
 			continue;
+		}
 
 		ret = fsl_spdif_txclk_caldiv(spdif_priv, clk, savesub, index,
 					     i == STC_TXCLK_SPDIF_ROOT);
-		if (savesub == ret)
+		if (savesub == ret) {
+			devm_clk_put(&pdev->dev, clk);
 			continue;
+		}
+
+		if (spdif_priv->txclk[index])
+			devm_clk_put(&pdev->dev, spdif_priv->txclk[index]);
 
 		savesub = ret;
 		spdif_priv->txclk[index] = clk;
-- 
2.7.4

